File: sfn/nested/SfnWorkflowInvocation.cs
using Newtonsoft.Json;

namespace sfn_ut.sfn.nested;

public class SfnWorkflowInvocation<PARENT,CHILD>(
  SfnTransform<PARENT, CHILD> transformInput,
  SfnWorkflow<CHILD> nested,
  SfnMerge<PARENT, CHILD> mergeOutput
  ) : SfnStepWithPayload<PARENT>(StepType.InvokeNested) {

  [JsonIgnore]
  public SfnTransform<PARENT, CHILD> TransformInput { get; private set; } = transformInput;
  [JsonIgnore]
  public SfnWorkflow<CHILD> Nested { get; private set; } = nested;
  [JsonIgnore]
  public SfnMerge<PARENT, CHILD> MergeOutput { get; private set; } = mergeOutput;

  public string NestedWorkflowName { get; private set; } = nested.WorkflowName;

  public override PARENT Execute(PARENT payload) {
    CHILD childIn = TransformInput.Execute(payload);
    CHILD childOut = Nested.Run(childIn);
    return MergeOutput.Execute(new(payload, childOut));
  }
}

File: java-sfn-unit-test/src/main/java/com/sfnut/sfn/nested/SfnWorkflowInvocation.java
package com.sfnut.sfn.nested;

import com.sfnut.sfn.ObjectMapping;
import com.sfnut.sfn.SfnStepWithPayload;
import com.sfnut.sfn.SfnWorkflow;
import com.sfnut.sfn.StepType;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;

public class SfnWorkflowInvocation<PARENT, CHILD> extends SfnStepWithPayload<PARENT> {
    @JsonIgnore
    private ObjectMapping<PARENT, CHILD> parentToChild;
    
    @JsonIgnore
    private SfnWorkflow<CHILD> nested;
    
    @JsonIgnore
    private ObjectMapping<CHILD, PARENT> childToParent;
    
    @JsonProperty("NestedWorkflowName")
    private String nestedWorkflowName;

    public SfnWorkflowInvocation(ObjectMapping<PARENT, CHILD> parentToChild,
                                SfnWorkflow<CHILD> nested,
                                ObjectMapping<CHILD, PARENT> childToParent) {
        super(StepType.INVOKE_NESTED);
        this.parentToChild = parentToChild;
        this.nested = nested;
        this.childToParent = childToParent;
        this.nestedWorkflowName = nested.getWorkflowName();
    }

    // Default constructor for Jackson
    public SfnWorkflowInvocation() {
        super(StepType.INVOKE_NESTED);
    }

    @Override
    public PARENT execute(PARENT payload) {
        CHILD childIn = parentToChild.map(payload);
        CHILD childOut = nested.run(childIn);
        return childToParent.mergeInto(childOut, payload);
    }

    public ObjectMapping<PARENT, CHILD> getParentToChild() {
        return parentToChild;
    }

    public SfnWorkflow<CHILD> getNested() {
        return nested;
    }

    public ObjectMapping<CHILD, PARENT> getChildToParent() {
        return childToParent;
    }

    public String getNestedWorkflowName() {
        return nestedWorkflowName;
    }
}
